### PR TITLE
Macsec scale test on a t2 topo

### PR DESCRIPTION
**Pull request by judyjoseph**
Summary:
Fixes # 13244
Add a new testcase to cover the link down and rekey in case of a linecard with large number of macsec sessions across multiple asics.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
Add a new testcase to cover the link down and rekey in case of a linecard with large number of macsec sessions across multiple asics.

#### How did you verify/test it?
Ran thescale test get pass on a testbed with LC having 32 macsec sessions.

```
jujoseph@sonic-mgmt-jujoseph_master:/var/src/sonic-mgmt-int/tests$ ./run_tests.sh -u -n vmsvc3-t2-xxxx-1 -c macsec/test_deployment.py::**TestDeployment::test_scale_rekey** -f ../ansible/testbed.yaml -i ../ansible/strsvc,../ansible/veos -t 't2,any' -e "--skip_sanity --disable_loganalyzer -x --enable_macsec --macsec_profile=MACSEC_PROFILE"                                                                                                                                      
=== Running tests in groups ===
Running: python2 -m pytest macsec/test_deployment.py::TestDeployment::test_scale_rekey --inventory ../ansible/strsvc,../ansible/veos --host-pattern svcstr-xxxx-lc1-1,svcstr-xxxx-lc2-1,svcstr-xxxx-lc3-1,svcstr-xxxx-sup-1 --testbed vmsvc3-t2-xxxx-1 --testbed_file ../ansible/testbed.yaml --log-cli-level warning --log-file-level debug --kube_master unset --showlocals --assert plain --show-capture no -rav --allow_recover --ignore=ptftests --ignore=acstests --ignore=saitests --ignore=scripts --ignore=k8s --ignore=sai_qualify --junit-xml=logs/tr.xml --log-file=logs/test.log --topology t2,any --skip_sanity --disable_loganalyzer -x --enable_macsec --macsec_profile=MACSEC_PROFILE
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
=================================================================================== test session starts ====================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
ansible: 2.8.20
rootdir: /var/src/sonic-mgmt-int/tests, inifile: pytest.ini
plugins: ansible-2.2.4, celery-4.4.7, html-1.22.1, repeat-0.9.1, forked-1.3.0, metadata-1.11.0, xdist-1.28.0, allure-pytest-2.8.22
collecting ... 
----------------------------------------------------------------------------------- live log collection ------------------------------------------------------------------------------------
00:28:17 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:28:17 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:28:17 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:28:28 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - ptp-256
00:28:28 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - ptp-256
00:29:03 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:29:03 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:29:03 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - dpu
00:29:14 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - ptp-256
00:29:14 testbed.get_testbed_type                 L0261 WARNING| Unsupported testbed type - ptp-256
             PASSED                                                                                                   [100%] 

-------------------------------------------------------------- generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml ---------------------------------------------------------------
=============================================================================== 1 passed in 4292.18 seconds ==============================================================================
```

### CODE REVIEW

**Review by Pterosaur — APPROVED** [review]

**Comment by judyjoseph** [comment]
@StormLiangMS could you please help review and merge this PR

File: tests/macsec/test_deployment.py [hunk 1]
```diff
@@ -23,3 +26,45 @@ def test_config_reload(self, duthost, ctrl_links, policy, cipher_suite, send_sci
         assert wait_until(300, 6, 12, check_appl_db, duthost, ctrl_links, policy, cipher_suite, send_sci)
         # Recover the original config file
         duthost.shell("sudo cp config_db.json /etc/sonic/config_db.json")
+
+    @pytest.mark.disable_loganalyzer
+    def test_scale_rekey(self, duthost, ctrl_links, rekey_period, wait_mka_establish):
```
Comment (arlakshm) [inline tests/macsec/test_deployment.py:31]: duthost will also be for the uplink card?

File: tests/macsec/test_deployment.py [hunk 2]
```diff
@@ -23,3 +26,45 @@ def test_config_reload(self, duthost, ctrl_links, policy, cipher_suite, send_sci
...
+        dut_ingress_sa_table_current = {}
+        new_dut_egress_sa_table = {}
+        new_dut_ingress_sa_table = {}
+
+        # Shut the interface and wait for all macsec sessions to be down
+        for dut_port, nbr in ctrl_links.items():
+            _, _, _, dut_egress_sa_table_orig[dut_port], dut_ingress_sa_table_orig[dut_port] = get_appl_db(
+                duthost, dut_port, nbr["host"], nbr["port"])
+            intf_asic = duthost.get_port_asic_instance(dut_port)
+            duthost.shell("config interface {} shutdown {}".format(intf_asic.cli_ns_option, dut_port))
```
Comment (arlakshm) [inline tests/macsec/test_deployment.py:44]: use the function in sonic_asic.py iso of CLI commands here https://github.com/sonic-net/sonic-mgmt/blob/aa1d42a81941fda710310a4b46a33c6c71684741/tests/common/devices/sonic_asic.py#L535

File: tests/macsec/test_deployment.py [hunk 3]
```diff
@@ -23,3 +26,45 @@ def test_config_reload(self, duthost, ctrl_links, policy, cipher_suite, send_sci
...
+
+        sleep(TestDeployment.MKA_TIMEOUT)
+
+        # Unshut the interfaces so that macsec sessions come back up
+        for dut_port, nbr in ctrl_links.items():
+            intf_asic = duthost.get_port_asic_instance(dut_port)
+            duthost.shell("config interface {} startup {}".format(intf_asic.cli_ns_option, dut_port))
```
Comment (arlakshm) [inline tests/macsec/test_deployment.py:51]: use the function in sonic_asic.py https://github.com/sonic-net/sonic-mgmt/blob/aa1d42a81941fda710310a4b46a33c6c71684741/tests/common/devices/sonic_asic.py#L531 iso CLI commands here

Comment (judyjoseph) [reply tests/macsec/test_deployment.py]: done

File: tests/macsec/test_deployment.py (hunk 1 above)
Comment (judyjoseph) [reply tests/macsec/test_deployment.py:31]: Currently it takes the first card (uplink) by default for all macsec tests, I have a follow up PR to fix all . plan to use the fixture macsec_duthost.

**Review by arlakshm — APPROVED** [review]

**Pull request merged**